feat(Scylla): add extension for scyllacomics.xyz

LucifersCircle opened the pull request:
add an extension for scyllacomics.xyz

LucifersCircle committed:
Add Scylla

add extension for scyllacomics.xyz

celarye commented on src/Scylla/main.ts:
```diff
@@ -0,0 +1,367 @@
+type Metadata = {
+    page?: number;
+    completed?: boolean;
+};
```
Move this to `models.ts`:
```suggestion

celarye commented on src/Scylla/main.ts:
```diff
@@ -0,0 +1,367 @@
+    ): Promise<ArrayBuffer> {
+        return data;
+    }
+}
```
Move this to `interceptors.ts`:
```suggestion

```

celarye commented on src/ScyllaComics/main.ts:
```diff
@@ -0,0 +1,367 @@
+        storage: "stateManager",
+    });
+
+    async initialise(): Promise<void> {
```
Note that interceptors get called in the order they are registered, this means you probably want to put rate limiter last as it blocks the two other interceptors as well.

celarye commented on src/ScyllaComics/main.ts:
```diff
@@ -0,0 +1,367 @@
+                title: "Most Viewed",
+                type: DiscoverSectionType.prominentCarousel,
+            },
+            //{
```
Getting it to work nicely with Paperback should be possible as well, we can look into this for a future release.

LucifersCircle replied:
Future release!

celarye commented on src/Scylla/main.ts:
```diff
@@ -0,0 +1,367 @@
+                cookie.name.startsWith("__cf")
+            ) {
+                // Find existing cookie with the same name
+                const existingCookie =
```
I know you probably copied this over from somewhere else but I am wondering if we really need to look for and delete the old cookie before setting the new one?

celarye commented on src/Scylla/main.ts:
```diff
@@ -0,0 +1,367 @@
+    MangaProviding,
+    PagedResults,
+    PaperbackInterceptor,
+    URL as PBURL,
```
Why the rename?

LucifersCircle replied:
I kept getting some sort of mix up with a vs code config or node.js where it kept trying to use node.js URL when I'd do url = new URL. I opted to try an alias so it always passed through types, and it worked alright, plus made the code easier on the eyes. Not sure what I could do to sort this out properly but could take a look if really necessary.

celarye replied:
Please take another look, when typing `new URL` make sure you do not accidentially import Node's `URL`.

celarye commented on src/ScyllaComics/main.ts:
```diff
@@ -0,0 +1,367 @@
+            .setQueryItem("status", "")
+            .setQueryItem("page", String(page));
+
+        const getFilterValue = (id: string) =>
+            query.filters?.find((filter) => filter.id === id)?.value;
+
+        const genres =
+            (getFilterValue("genres") as Record<
+                string,
+                "included" | "excluded"
+            >) ?? {};
+
+        for (const [key, value] of Object.entries(genres)) {
+            if (value === "included") {
+                url.setQueryItem("genre[]", key);
+            }
+        }
+
+        // strict matching filter, look into getting paperback to use this with a toggle
+        const strictVal = getFilterValue("strict");
+        const isStrict =
+            typeof strictVal === "string" &&
+            (strictVal.toLowerCase() === "on" ||
+                strictVal.toLowerCase() === "true");
+        if (isStrict) {
+            url.setQueryItem("strict", "on");
+        }
```
I would personally move the Paperback filter parsing into its own method:
```suggestion
        const getFilterValue = (id: string) =>
            query.filters?.find((filter) => filter.id === id)?.value;

        const genres =
            (getFilterValue("genres") as Record<
                string,
                "included" | "excluded"
            >) ?? {};

        for (const [key, value] of Object.entries(genres)) {
            if (value === "included") {
                url.setQueryItem("genre[]", key);
            }
        }

        // strict matching filter, look into getting paperback to use this with a toggle
        const strictVal = getFilterValue("strict");
        const isStrict =
            typeof strictVal === "string" &&
            (strictVal.toLowerCase() === "on" ||
                strictVal.toLowerCase() === "true");
        if (isStrict) {
            url.setQueryItem("strict", "on");
        }
```

celarye commented on src/Scylla/main.ts:
```diff
@@ -0,0 +1,367 @@
+        }
+        if (status === 403) {
+            throw new Error(
+                "Server returned 403 Forbidden. Logins not yet supported. Let me know on Discord if you can view this title when logged in on the website.",
```
Users only have the contact information you list in the developer section of the `pbconfig.ts` file, you probably want to be more descriptive with your contact information. Asking them to make a support post in the Inkdex extension repository on GitHub or Discord server would be better imo (links are not interactive or copyable in Paperback).
```suggestion
                "Status code: 403 Forbidden. Login is not yet supported. Make a support issue/post in the Inkdex extension repository on their GitHub Organization or Discord server if you can view this title when logged in on the website.",
```

LucifersCircle replied:
I at first believed it was for users who could login only, however I joined their discord to ask about it and they just send 403 responses for title pages that got taken down via DMCA, and for some reason leave the titles listed. I updated the comment to reflect this.

celarye commented on src/Scylla/pbconfig.ts:
```diff
@@ -0,0 +1,24 @@
+    version: "1.0.0-alpha.1",
+    icon: "icon.png",
+    language: "en",
+    contentRating: ContentRating.ADULT,
```
This should reflect the majority content rating for titles on the source.

celarye commented on src/Scylla/pbconfig.ts:
```diff
@@ -0,0 +1,24 @@
+        {
+            name: "Lucifers Circle",
+            website: "https://inkdex.github.io",
+            github: "https://github.com/inkdex",
```
Remove these or put your personal ones:
```suggestion
```

celarye commented on src/Scylla/parsers.ts:
```diff
@@ -0,0 +1,360 @@
+        const parentLink = $(obj).attr("href") ?? "";
+        const chapterId = parentLink.replace(/\/$/, "").split("/").pop() ?? "";
+        if (!chapterId || chapterId === "#") {
+            console.log(
```
I am not sure console logging these is enough? This would maybe go unnoticed to the user and cause confusion. While throwing an error will prevent reading all together it will cause them to report an actuall issue needing fixing.

celarye commented on src/Scylla/parsers.ts:
```diff
@@ -0,0 +1,360 @@
+        const num = parseInt(match[1], 10);
+        const unit = match[2];
+
+        const multipliers: Record<string, number> = {
```
Move this to `models.ts`:
```suggestion
```

LucifersCircle committed:
Update Scylla Comics extension

chore(Scylla): move Metadata type to models.ts
chore(Scylla): move parseRelativeDate multipliers to models.ts
chore(Scylla): move ScyllaInterceptor to interceptors.ts
chore(Scylla): reorder interceptor registration
chore(Scylla): simplify saveCloudflareBypassCookies
chore(Scylla): extract search filter parsing into helper
chore(Scylla): set contentRating to MATURE to reflect majority of titles
fix(Scylla): update checkCloudflareStatus 403 handling (DMCA notice instead of login)
fix(Scylla): throw error on missing chapterId instead of console log
docs(Scylla): add Scylla Comics to README

celarye commented on src/Scylla/interceptors.ts:
```diff
@@ -0,0 +1,23 @@
+            ...(request.headers ?? {}),
+            ...{
+                referer: `${SCYLLA_DOMAIN}/`,
+                "user-agent": await Application.getDefaultUserAgent(),
+            },
+        };
+        return request;
+    }
```
Setting the user agent to the default should not be necessary
```suggestion
    async interceptRequest(request: Request): Promise<Request> {
        request.headers.referer = `${SCYLLA_DOMAIN}/`;

        return request;
    }
```

LucifersCircle committed:
Remove PBURL alias

chore(Scylla): replace PBURL alias with correct URL

LucifersCircle committed:
Update pbconfig.ts

chore(Scylla): update contentRating to ADULT

LucifersCircle committed:
Update interceptors.ts

refactor(Scylla): remove user agent from request interceptor

LucifersCircle committed:
re-run prettier

style(Scylla): format code using Prettier

celarye commented on src/Scylla/parsers.ts:
```diff
@@ -0,0 +1,330 @@
+import { CheerioAPI } from "cheerio";
+import { TIME_MULTIPLIERS } from "./models";
+
+export const parseMangaDetails = (
```
Change all parsers to normal functions.

LucifersCircle committed:
Update parsers.ts

refactor(Scylla): convert parser exports to standard function declarations

LucifersCircle committed:
Change Scylla to ScyllaComics

refactor(ScyllaComics): rename all Scylla references to ScyllaComics
